src/event.rs

```diff
@@ -0,0 +1,622 @@
+				Ok(Self::PaymentSuccessful { payment_hash })
+			}
+			1u8 => {
+				let payment_hash: PaymentHash = Readable::read(reader)?;
+				Ok(Self::PaymentFailed { payment_hash })
+			}
+			2u8 => {
+				let payment_hash: PaymentHash = Readable::read(reader)?;
+				let amount_msat: u64 = Readable::read(reader)?;
+				Ok(Self::PaymentReceived { payment_hash, amount_msat })
+			}
+			3u8 => {
+				let channel_id: [u8; 32] = Readable::read(reader)?;
+				let user_channel_id: u64 = Readable::read(reader)?;
+				Ok(Self::ChannelReady { channel_id, user_channel_id })
+			}
+			4u8 => {
+				let channel_id: [u8; 32] = Readable::read(reader)?;
+				let user_channel_id: u64 = Readable::read(reader)?;
+				Ok(Self::ChannelClosed { channel_id, user_channel_id })
+			}
+			//5u8 => {
+			// TODO OnChainPaymentSent
+			//}
+			//6u8 => {
+			// TODO OnChainPaymentReceived
+			//}
+			_ => Err(lightning::ln::msgs::DecodeError::InvalidValue),
+		}
+	}
+}
+
+impl Writeable for LdkLiteEvent {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), lightning::io::Error> {
+		match self {
+			Self::PaymentSuccessful { payment_hash } => {
+				0u8.write(writer)?;
+				payment_hash.write(writer)?;
+				Ok(())
+			}
+			Self::PaymentFailed { payment_hash } => {
+				1u8.write(writer)?;
+				payment_hash.write(writer)?;
+				Ok(())
+			}
+			Self::PaymentReceived { payment_hash, amount_msat } => {
+				2u8.write(writer)?;
+				payment_hash.write(writer)?;
+				amount_msat.write(writer)?;
+				Ok(())
+			}
+			Self::ChannelReady { channel_id, user_channel_id } => {
+				3u8.write(writer)?;
+				channel_id.write(writer)?;
+				user_channel_id.write(writer)?;
+				Ok(())
+			}
+			Self::ChannelClosed { channel_id, user_channel_id } => {
+				4u8.write(writer)?;
+				channel_id.write(writer)?;
+				user_channel_id.write(writer)?;
+				Ok(())
+			} //Self::OnChainPaymentSent { .. } => {
+			  //TODO
+			  //}
+			  //Self::OnChainPaymentReceived { .. } => {
+			  //TODO
+			  //}
+		}
+	}
+}
+
+pub(crate) struct LdkLiteEventQueue<K: Deref>
+where
+	K::Target: KVStorePersister,
+{
+	queue: Mutex<VecDeque<Arc<Event>>>,
+	notifier: Condvar,
+	persister: K,
+}
+
+impl<K: Deref> LdkLiteEventQueue<K>
+where
+	K::Target: KVStorePersister,
+{
+	pub(crate) fn new(persister: K) -> Self {
+		let queue: Mutex<VecDeque<Arc<Event>>> = Mutex::new(VecDeque::new());
+		let notifier = Condvar::new();
+		Self { queue, notifier, persister }
+	}
+	pub(crate) fn add_event(&self, event: LdkLiteEvent) -> Result<(), Error> {
+		{
+			let mut locked_queue = self.queue.lock().unwrap();
+			locked_queue.push_back(Arc::new(event));
+			self.persist_queue(&*locked_queue)?;
+		}
+
+		self.notifier.notify_one();
+		Ok(())
+	}
+
+	pub(crate) fn next_event(&self) -> Arc<LdkLiteEvent> {
+		let locked_queue = self
+			.notifier
+			.wait_while(self.queue.lock().unwrap(), |queue| queue.is_empty())
+			.unwrap();
+		Arc::clone(&locked_queue.front().unwrap())
+	}
```

Comment:
Curious how this would work with shutdown. If a user has a loop calling `next_event`, do they need to wait for an event comes through to exit the loop? Would using `wait_timeout_while` be preferable?

Reply:
Hm, yeah, that is a good question regarding how we imagine the overall API to be used (i.e., if we expect the user side to entirely run single-threaded or if there would be a background process blocking on this). I'm not the biggest fan of timeouts, but maybe we should expose multiple versions of this via the main API? One blocking, one blocking for a given timeout, and one returning `Future`?